[AlwaysOnProfiler] Memory profiler internal docs (#641)
signalfx/signalfx-dotnet-tracing: docs/internal/memory-profiling.md

> # About the .NET Memory Profiling

Reviewer: ```suggestion
# About the AlwaysOn memory profiling for .NET
```

> The SignalFx Instrumentation for .NET includes a Memory Profiler
> that can be enabled with a configuration setting. This profiler samples allocations,
> captures the call stack state for .NET thread that triggered the allocation,
> and sends these to the Splunk Observability Cloud as logs.
> 
> Memory allocation data, together with the stack traces and dotnet runtime metrics,
> can then be used to investigate memory leaks and unusual consumption patterns.

Reviewer: ```suggestion
The SignalFx Instrumentation for .NET includes a memory profiler for Splunk
APM that can be enabled with a setting. The profiler samples allocations,
captures the call stack state for the .NET thread that triggered the allocation,
and sends the telemetry to Splunk Observability Cloud.

Use the memory allocation data, together with the stack traces and .NET runtime metrics,
to investigate memory leaks and unusual consumption patterns in AlwaysOn Profiling.
```

The fact that we send this data as logs, AFAIK, is just an implementation detail. It's not relevant to users.

> to investigate memory leaks and unusual consumption patterns in AlwaysOn Profiling.

Reviewer 2: To me this reads like one is investigating memory leaks and consumption in "AlwaysOn Profiling". I will defer to @theletterf judgement here.

Author: Current wording was suggested in https://github.com/signalfx/signalfx-dotnet-tracing/pull/641/commits/a6d5fe2c59773f56740829510eab4423b7b0dcb0

## How does the memory profiler work?

The profiler leverages [.NET profiling](https://docs.microsoft.com/en-us/dotnet/framework/unmanaged-api/profiling/)
to perform allocation sampling.

> For every sampled allocation, allocation amount together with state of the thread

Reviewer 3: ```suggestion
For every sampled allocation, allocation amount as well as the state of the thread
```

> For every sampled allocation, allocation amount as well as the state of the thread

Reviewer 2: The state of the thread is always the same, no? I mean it is a running thread. I understand that this is part of the export format info but seems redundant here.

Author: reworded to avoid confusion in https://github.com/signalfx/signalfx-dotnet-tracing/pull/641/commits/62ed5a18c668cea5418b3c08b41aa0547c0f0bfb

that triggered the allocation are saved into buffer.

> The managed-thread shared with [CPU Profiler](../always-on-profiling.md)
> processes the data from the buffer and sends it to the OpenTelemetry Collector.

Reviewer: ```suggestion
The managed thread shared with [CPU Profiler](../always-on-profiling.md)
processes the data from the buffer and sends it to the OpenTelemetry Collector.
```

Stack trace data is embedded as a string inside of an OTLP logs payload. The
[Splunk OpenTelemetry Collector](https://github.com/signalfx/splunk-otel-collector)
detects profiling data inside OTLP logs and forwards it to
Splunk APM.

# Requirements

> * .NET 5.0 or higher (`ICorProfilerInfo12` available in runtime).

Reviewer 4: The question is if we should reference in memory profiling .NET 5 or 6. 
.NET 5 is EOL.

Reviewer 2: Since our current requirements say .NET 5.0 (see https://github.com/signalfx/signalfx-dotnet-tracing#requirements) I vote for .NET 5.0

* [Splunk OpenTelemetry Collector](https://github.com/signalfx/splunk-otel-collector)
version 0.34.0 or higher.
_Sending profiling data directly to ingest is not supported at this time_.

# Enable the profiler

To enable the profiler, set the `SIGNALFX_PROFILER_MEMORY_ENABLED` environment variable
to `true` for your .NET process.

# Configuration settings

> Please check the descriptions for the following environment variable:

Reviewer: ```suggestion
Make sure you're following the documentation for the following environment variables:
```

* [`SIGNALFX_PROFILER_MEMORY_ENABLED`](../internal/internal-config.md#internal-settings)
* [`SIGNALFX_PROFILER_LOGS_ENDPOINT`](../advanced-config.md#alwayson-profiling-settings)

> > _NOTE_: `SIGNALFX_PROFILER_LOGS_ENDPOINT` affects both CPU and Memory profiling.

Reviewer 4: Do we need same note for always-on-profiler?

Author: added in https://github.com/signalfx/signalfx-dotnet-tracing/pull/641/commits/6b90d6bcbacf338bb3bfe01d45f532a6a5c1fc75

# Escape hatch

The profiler limits its own behavior when buffer

> used to store allocation samples is full.

Author: added limit information in https://github.com/signalfx/signalfx-dotnet-tracing/pull/641/commits/62ed5a18c668cea5418b3c08b41aa0547c0f0bfb
